File: Surrounded-Regions/Surrounded-Regions.cpp

class Solution {
public:
    int n;
    int m;
    int dx[4]={0,1,0,-1};
    int dy[4]={1,0,-1,0};
    bool ok(int x,int y)
    {
        return x>=0 && x<n && y>=0 && y<m;
    }
    void bfs(vector<vector<char> >&board,int x,int y)
    {
        queue<int>qx;
        queue<int>qy;
        qx.push(x);
        qy.push(y);
        board[x][y]='A';
        while(!qx.empty())
        {
            int xx=qx.front();
            qx.pop();
            int yy=qy.front();
            qy.pop();
            for(int i=0;i<4;i++)
            {
                int tx=xx+dx[i];
                int ty=yy+dy[i];
                if(ok(tx,ty) && board[tx][ty]=='O')
                {
                    qx.push(tx);
                    qy.push(ty);
                    board[tx][ty]='A';
                }
            }
        }
    }
    void solve(vector<vector<char>> &board) {
        n=board.size();
        if(n==0)
            return;
        m=board[0].size();
        for(int i=0;i<n;i++)
        {
            if(board[i][0]=='O')
                bfs(board,i,0);
            if(board[i][m-1]=='O')
                bfs(board,i,m-1);
        }
        for(int j=0;j<m;j++)
        {
            if(board[0][j]=='O')
                bfs(board,0,j);
            if(board[n-1][j]=='O')
                bfs(board,n-1,j);
        }
        for(int i=0;i<n;i++)
            for(int j=0;j<m;j++)
            {
                if(board[i][j]=='O')
                    board[i][j]='X';
                else if(board[i][j]=='A')
                    board[i][j]='O';
            }
    }
};


 

File: solutions/Surrounded-Regions/Surrounded-Regions.py


class Solution:
    # @param board, a 2D array
    # Capture all regions by modifying the input board in-place.
    # Do not return any value.
    def solve(self, board):
        n=len(board)
        if n==0:
            return
        m=len(board[0])
        if m==0:
            return
        for i in range(n):
            if board[i][0]=='O':
                self.bfs(board,i,0)
            if board[i][m-1]=='O':
                self.bfs(board,i,m-1)
        for j in range(m):
            if board[0][j]=='O':
                self.bfs(board,0,j)
            if board[n-1][j]=='O':
                self.bfs(board,n-1,j)
        for i in range(n):
            for j in range(m):
                if board[i][j]=='O':
                    board[i][j]='X'
                elif board[i][j]=='A':
                    board[i][j]='O'
    def bfs(self,board,x,y):
        dx = []
        dx = [0,1,0,-1]
        dy = []
        dy = [1,0,-1,0]
        n=len(board)
        m=len(board[0])
        qx = collections.deque()
        qy = collections.deque()
        qx.append(x)
        qy.append(y)
        board[x][y]='A'
        while qx:
            xx=qx.popleft()
            yy=qy.popleft()
            for i in range(4):
                tx=xx+dx[i]
                ty=yy+dy[i]
                if self.ok(tx,ty,n,m) and board[tx][ty]=='O':
                    qx.append(tx)
                    qy.append(ty)
                    board[tx][ty]='A'
    def ok(self,x,y,n,m):
        return x>=0 and x<n and y>=0 and y<m
 